### PR TITLE
Improve usability of the DriverAvailabilities Create Page

### LINKED ISSUE #13: EPIC: Improve usability of the DriverAvailabilities Create Page
_### As a...

driver

### I can...

Undestand what each of the form fields mean

### So that...

i can submit a correct availability without running into issues

### Discussion

The current create page for DriverAvailabilities has a number of aspects that should be improved.

First:
When create an availability we do not need to take the driver id as input, rather it should be automatically assigned to the creator.

Second:
image
The form labels are very unprofessional and jarring. We need to change this to be consistent with the rest of the application. Please capitalize and separate words.

Third:
image
Input validation error messages on bad input also must be improved with capitalization and spaces.

Fourth:
image
There is no 'bad-input' error messages. We should have error messages on bad input for `Drive Id`, `Day`, `Start Time`, `End Time`. For day it would be ideal to convert this field to a dropdown, similar to the one on the `RideRequestForm`.

Fifth
Like the `RideRequestForm` we want to have example messages on each of the form fields to help making it easier for the driver to know how to fill it out
Example:
image

### Acceptance Criteria

- [x] The form labels are improved
- [x] The bad input and missing input error messages are added and improved
- [x] Example inputs on each form field

### Implementation Todos

Issue 1 : Backend:
- [x] Modify `DriverAvailabilityController.java` `POST` so that when creating a new `DriverAvailability` we automatically use the creator's id and no longer take in id as an input.
- [x] Modify `DriverAvailabilityController.java` `PUT` so that when editing `DriverAvailability` we make sure that the `DriverAvailability` we intend to edit is really owned by the current user, and we no longer want to set the driver id to the incoming.
- [x] Add the proper validations to `GET` single and `DELETE` so that a user can only delete or get their own availabilities and not anyone elses

Issue 2 : Frontend:
- [x] Modify `DriverAvailabilityForm.js` to meet acceptance criterea.

You may further split this into more issues if deemed necessary

### PR DESCRIPTION
**Pull request by Ononymous**
# This branch depends on branch #45  

### As a...

driver

### I can...

Undestand what each of the form fields mean

### So that...

i can submit a correct availability without running into issues

image

image

Create and Edit driver availability page required the input of driver id, which was not very user friendly.

The current page does not require that field for the POST request. Instead, the backend automatically adds the driver id in the post request to database.

Also, added a check such that the Get single, Delete, and Edit requests only allow availability entries of this specific user/driver.

Closes #13

### CODE REVIEW

**Review by Tommygithubaccount123 — DISMISSED** [review]
LGTM

Damn nice, I didn't know to use `getCurrentUser().getUser().getId();` for ID

**Review by Tommygithubaccount123 — CHANGES_REQUESTED** [review]
IMPORTANT:

I added changes to my branch, so you need to merge my changes into your branch!!!!
https://github.com/ucsb-cs156-s24/proj-gauchoride-s24-5pm-5/pull/29

**Comment by Ononymous** [comment]
> IMPORTANT:
> 
> I added changes to my branch, so you need to merge my changes into your branch!!!! #29

Merged conflicts and new changes from branch

**Review by Tommygithubaccount123 — DISMISSED** [review]
LGTM

**Comment by Ononymous** [comment]
@pconrad Added deployed link, with 2 test availabilities inputted

**Comment by Ononymous** [comment]
@pconrad 
Sorry about the extra step. The environment variable has been fixed. It seems that I have repeated the word "ADMIN_EMAILS=" twice, resulting in professor's email to be invalid.

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverAvailabilityController.java [hunk 1]
```diff
@@ -120,6 +130,10 @@ public Object deleteDriverAvailability(
         DriverAvailability availability = driverAvailabilityRepository.findById(id)
                 .orElseThrow(() -> new EntityNotFoundException(DriverAvailability.class, id));
 
+        if (availability.getDriverId() != getCurrentUser().getUser().getId()) {
+            throw new EntityNotFoundException(DriverAvailability.class, id);
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverAvailabilityController.java]: This is not the correct exception to throw.  This should only be used when you did a search for a DriverAvailability with a certain id and it was not found.  Throwing this exception will cause future devs to curse your name.

The exception you want is one that will turn into a 403 unauthorized, but if you can't figure out how to do that, then just throw a generic exception with a reasonable error message.

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverAvailabilityController.java [hunk 2]
```diff
@@ -95,12 +104,13 @@ public ResponseEntity<Object> updateDriverAvailability(
                             @RequestParam Long id,
                             @RequestBody @Valid DriverAvailability incoming)
     {
-        DriverAvailability availability;
+        DriverAvailability availability = driverAvailabilityRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(DriverAvailability.class, id));
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverAvailabilityController.java:110]: This is the correct use of this exception,

File: src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverAvailabilityController.java [hunk 3]
```diff
@@ -95,12 +104,13 @@ public ResponseEntity<Object> updateDriverAvailability(
...
 
-        availability = driverAvailabilityRepository.findById(id)
-                    .orElseThrow(() -> new EntityNotFoundException(DriverAvailability.class, id));
+        if (availability.getDriverId() != getCurrentUser().getUser().getId()) {
```
Comment (pconrad) [inline src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverAvailabilityController.java:112]: Member
@[pconrad](https://github.com/pconrad) pconrad [Jun 4, 2024](https://github.com/ucsb-cs156-s24/proj-gauchoride-s24-5pm-5/pull/37/files#r1626804244)

This is not the correct exception to throw. This should only be used when you did a search for a DriverAvailability with a certain id and it was not found. Throwing this exception will cause future devs to curse your name.

The exception you want is one that will turn into a 403 unauthorized, but if you can't figure out how to do that, then just throw a generic exception with a reasonable error message.
@pconrad

**Review by pconrad — CHANGES_REQUESTED** [review]
This code is throwing a misleading exception.

Please fix; otherwise this is not mergable.  You have only until I finish with other team's PRs, or the end of section tomorrow night, which ever is sooner.

**Comment by Ononymous** [comment]
> This code is throwing a misleading exception.
> 
> Please fix; otherwise this is not mergable. You have only until I finish with other team's PRs, or the end of section tomorrow night, which ever is sooner.

@pconrad Hi Professor, I just pushed the new commit that modified the exception behavior for 403 forbidden in Driver Availability.

I temporarily added the fixes from PR #45 , and it is working in swaggers when I try edit/delete/get a driver availability that doesn't belong to the user.

image

**Review by rileychou — DISMISSED** [review]
LGTM, nice fix on the Forbidden error messages.

**Review by pconrad — CHANGES_REQUESTED** [review]
I don't see any problems with the code, per se, but when I go to test this, I'm just confused.  There are two shift menus on the navbar, and I'm not sure which I should be testing.  One of them has backend errors showing, up, and the other doesn't seem to match the PR.

Can you make a more clear test plan?  And consider cleaning up the UI issues that make it confusing to use and test?

**Review by pconrad — DISMISSED** [review]
LGTM

**Pull request merged**